src/compiler/binder.ts

```diff
@@ -48,12 +48,16 @@ namespace ts {
                 }
                 break;
             // 3. non-exported import declarations
-            case SyntaxKind.ImportDeclaration:
             case SyntaxKind.ImportEqualsDeclaration:
-                if (!(hasModifier(node, ModifierFlags.Export))) {
-                    return ModuleInstanceState.NonInstantiated;
+                const container = findAncestor(node, (n): n is ModuleDeclaration | SourceFile => isModuleDeclaration(n) || isSourceFile(n));
+                if (container && isExportedInBlock(idText((node as ImportEqualsDeclaration).name), container)) {

```

Comment:
I'm pretty sure `getModuleInstanceStateForExportedAlias` is supposed to handle this, as it's support to start at the export, follow it to the import, then check and return the state of the import target. At least, I think that's a better place to handle it - is already does a full scope traversal looking for things, and handles circularities via the cache layer on the instantiation state, it just doesn't handle import declarations (just module declarations).

Reply:
Specifically, I think this, as is, is backwards. The act of exporting the import is _not_ (generally speaking) what makes a namespace "instantiated" - the namespace exporting some value is, and the default state, lacking good heuristics, is to assume that it might export a value. My contention is mostly that if you write
```ts
namespace Foo {
  export namespace Bar {}
}
namespace Baz {
  import B = Foo.Bar;
  export { B }
}
```
`Baz` should _not_ be instantiated because `Foo.Bar` isn't instantiated. While if you write
```ts
namespace Foo {
  export namespace Bar {
    export class Cls {}
  }
}
namespace Baz {
  import B = Foo.Bar;
  export { B }
}
```
it should be, because now `Foo.Bar` contains a non-type-only declaration, thus is instantiated, thus a namespace which reexports it is instantiated.